nixosMinimal: init - nixos with minimal modules

DavHau opened the pull request:
related: https://github.com/NixOS/nixpkgs/issues/137168

This adds:
- a `minimal.nix` file containing a list of 185 modules that are required for building the `toplevel` of an (almost) empty nixos configuration.
- a test that builds a minimal nixos system to ensure that the minimal.nix doesn't miss any modules.

This is only the first step and further work is required in order to improve the experience using nixosMinimal, like:
- refactoring existing modules to explicitly import their dependency modules
- adding CI checks to assure modules declare imports correctly

cc @roberth 

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

roberth commented:
I'm concerned about the process and the seemingly arbitrary nature of the package set here.

For instance `nixpkgs/misc.nix` has a smaller alternative in the form of `nixosModules.readOnlyPkgs` in `flake.nix`, and in theory you don't even need a module for it at all.

I think a more "decidable" approach is to begin with the absolute minimum, which is `nixpkgsFlake.lib.nixos.evalModules`, and then only add the smallest common intersection of some very minimal use cases, while also reviewing whether any of the modules that are loaded are actually minimal for their purpose. Rather often, they are not, containing definitions that serve other modules or other use cases that are not relevant. Maybe it's ok to have `system.build.toplevel` even if some use cases don't need it, but I'm only 95% convinced.
An example of a minimal use case would be to generate a docker image with a NixOS-derived `/etc`, which doesn't need most of the modules listed. (I don't think that code ever made it into master, but you can find it in https://github.com/NixOS/nixpkgs/pull/148456)

Which goal are you pursuing?
It could be evaluation performance, in which case #372170 also helps - removing the bulk of `services` modules.
Or is it flexibility? E.g. upstreaming https://github.com/arianvp/server-optimised-nixos?

If flexibility is what you're going for, I believe the only viable, sustainable approach is to turn each module or small set of modules into a project of their own, which is to say, having their own tests that are independent of all the other modules except the ones they import, and `meta.maintainers` (which already exists for NixOS modules, but is almost never used).
Imagine that, actual modularity!

roberth commented:
Awesome that you want to work on this. I hope my comment was useful.

Lassulus commented:
The idea of this PR is to have a base operating system which can be used for different approaches. For example a more minimal nixos as in server-optimized-nixos or as a base platform for modular services.

Currently all nixos modules are automatically imported and we cannot change the default nixosSystem entry point to change that behavior without breaking it for all users. So our approach is to add a new entrypoint which can be used to build a bootable system. For now the list is big and arbitrary but with the changes made here, there is code and infrastructure so people can start asynchronously working on splitting up these modules. For example the pam module, which is uses a lot of other modules, instead of just declaring options and the users would use those options in return.

A next step would be, to track dependencies between modules with imports inside the modules, so we can show those dependencies explicitly. Hunting down these dependencies would also be easier with the new entrypoint.

All in all it's an iterative approach toward a modular services, but the goal here is to reuse as much of the current services we have already in nixos. This is the first minimal change, since minimal changes are easier to review :)

next steps would be something like:
- explicitly declare inter module dependencies with explicit imports
- add tests that all modules actually declare all their required imports
- refactor some modules to be more minimal and reduce the dependency list

roberth commented on nixos/modules/base-modules.nix:
This should be a module instead.

roberth commented on nixos/modules/base-modules.nix:
```diff
@@ -0,0 +1,187 @@
+[
```
This could be a module instead.

```suggestion
{ imports = [
```

roberth commented on nixos/tests/nixos-minimal.nix:
```diff
@@ -0,0 +1,11 @@
+  pkgs ? import ../.. { },
+}:
+let
+  nixos = pkgs.nixosMinimal {
```
```suggestion
  # aka `lib.nixos` in the nixpkgs flake
  nixos = import ../lib/default.nix { };
  conf = nixos.lib.evalModules { modules = [ {
```

roberth commented on pkgs/top-level/all-packages.nix:
```diff
@@ -18664,6 +18664,30 @@ with pkgs;
     in
     c.config.system.build // c;
 
+  nixosMinimal =
+    configuration:
+    let
+      c = import (path + "/nixos/lib/eval-config.nix") {
+        baseModules = import (path + "/nixos/modules/base-modules.nix");
+        modules = [
+          (
+            { lib, ... }:
+            {
+              config.nixpkgs.pkgs = lib.mkDefault pkgs;
+              config.nixpkgs.localSystem = lib.mkDefault stdenv.hostPlatform;
+              # TODO: fix documentation not building
+              config.documentation.enable = lib.mkDefault false;
+            }
+          )
+        ] ++ (if builtins.isList configuration then configuration else [ configuration ]);
+
+        # The system is inherited from the current pkgs above.
+        # Set it to null, to remove the "legacy" entrypoint's non-hermetic default.
+        system = null;
+      };
+    in
+    lib.warn "nixosMinimal is experimental and subject to change" c.config.system.build // c;
+
```
Use the existing minimal entrypoint instead.

```suggestion
  nixosMinimal =
    configuration:
    let
      c = import (path + "/nixos/lib/eval-config.nix") {
        baseModules = import (path + "/nixos/modules/base-modules.nix");
        modules = [
          (
            { lib, ... }:
            {
              config.nixpkgs.pkgs = lib.mkDefault pkgs;
              config.nixpkgs.localSystem = lib.mkDefault stdenv.hostPlatform;
              # TODO: fix documentation not building
              config.documentation.enable = lib.mkDefault false;
            }
          )
        ] ++ (if builtins.isList configuration then configuration else [ configuration ]);

        # The system is inherited from the current pkgs above.
        # Set it to null, to remove the "legacy" entrypoint's non-hermetic default.
        system = null;
      };
    in
    lib.warn "nixosMinimal is experimental and subject to change" c.config.system.build // c;

```

roberth reviewed (changes requested):
This should be done with less new stuff by leveraging the existing minimal entrypoint.

It'd be good to have a separation between the entrypoint and the module set (profile), so that we don't end up maintaining a ton of subtly distinct entrypoints. We already have too many.

DavHau committed:
nixosMinimal: init - nixos with minimal modules

related: https://github.com/NixOS/nixpkgs/issues/137168

This adds:
- a `minimal.nix` file containing a list of 185 modules that are required for building the `toplevel` of an (almost) empty nixos configuration.
- a test that builds a minimal nixos system to ensure that the minimal.nix doesn't miss any modules.

This is only the first step and further work is required in order to improve the experience using nixosMinimal, like:
- refactoring existing modules to explicitly import their dependency modules
- adding CI checks to assure modules declare imports correctly

DavHau commented:
@roberth I removed the additional entry point and converted the list to a module. Please check again

DavHau commented on nixos/modules/minimal.nix:
Is there a better location for this file?

roberth replied:
It's like a profile, an actually good one. Maybe `modules/profiles/minimal/baseline.nix` or something?

roberth commented on nixos/modules/minimal.nix:
```diff
@@ -0,0 +1,205 @@
+/*
+  This is a minimal list of nixos modules.
```
"a minimal" is still contradictory, unless we have multiple minimal sets of the same size. I don't believe that's the case.

I'd go with something less absolute and more vague like "core" or "sea level" or something, for the name.

roberth reviewed (commented):
Much better in terms of accidental complexity. Only remaining issue: how do we prevent this from bitrotting frequently?
Is this best-effort with async fixes, or is the plan to add CI for it?
